overlord/snapstate/handlers.go

```diff
@@ -1318,19 +1421,44 @@ func (m *SnapManager) doHomeMigration(t *state.Task, st *state.State, snapsup *S
 	return undo, nil
 }
 
-func (m *SnapManager) doFullMigration(t *state.Task, st *state.State, snapsup *SnapSetup, newInfo *snap.Info) (revert func(), err error) {
-	revertHidden, err := m.doHiddenMigration(t, st, snapsup, newInfo)
+func (m *SnapManager) revertFullMigration(t *state.Task, st *state.State, snapsup *SnapSetup, oldRev snap.Revision) (undo func(), err error) {
+	hiddenUndo, err := m.revertHiddenMigration(t, st, snapsup)
 	if err != nil {
-		return revertHidden, err
+		return hiddenUndo, err
 	}
 
-	revertHome, err := m.doHomeMigration(t, st, snapsup, newInfo.Revision)
+	homeUndo, err := m.revertHomeMigration(t, st, snapsup, oldRev)
 	return func() {
-		revertHidden()
-		revertHome()
+		homeUndo()
+		hiddenUndo()
 	}, err
 }
 
+func (m *SnapManager) revertHomeMigration(t *state.Task, st *state.State, snapsup *SnapSetup, oldRev snap.Revision) (undo func(), err error) {
+	undo = func() {
+		if err := m.backend.InitExposedSnapHome(snapsup.InstanceName(), oldRev); err != nil {
+			dir := filepath.Join("~", dirs.ExposedSnapHomeDir, snapsup.InstanceName())
+			st.Lock()
+			defer st.Unlock()
+			st.Warnf("cannot undo removal of %q: %v", dir, err)
+		}
+	}
+
+	if err := m.backend.RemoveExposedSnapHome(snapsup.InstanceName()); err != nil {
```

Comment:
sorry, I didn't think of this before but usually revert doesn't delete the data, but this is deleting part of the data of the new revision basically, no? not sure what it means

if a snap A is at revision  R, it's possible to do:

snap revert A
snap refresh --revision=R A

and the data that was there for revision R would come back as is afair

Reply:
Ok then I misunderstood what's required of revert. Ts the directory under ~/Snap meant to be saved somehow and then deleted or just not changed at all? The revert documentation states it discards "any data changes that were done by the latest revision" so maybe we should correct/expand that.